julia/src/api_types.jl

```diff
@@ -33,22 +33,35 @@ function bust_union(x::TaggedUnion{U}) where U
 end
 bust_union(x) = x  # not a union, leave it alone
 
+"Forces reinterpretting `raw` as a `T`. If sizeof(raw) <: size(T) high bits will be filled with junk."
 function force_reinterpret(::Type{T}, raw) where T
     isbits(raw) || throw(ArgumentError("Can only reinterpret from a isbits type"))
     isbitstype(T) || throw(ArgumentError("Can only reinterpret into a isbits type"))
     return unsafe_load(Ptr{T}(Base.pointer_from_objref(Ref(raw))))
 end
 
 
-struct CLit <: TaggedUnion{Tuple{Int64, Int32, Int8, Float64, Float32, UInt64, UInt32}}
+struct CLit <: TaggedUnion{Tuple{Int64, Int32, Int8, Float64, Float32, UInt32, UInt64}}
     tag::UInt64
     payload::NTuple{8, UInt8}  # actually the Union, needs to be big enough to hold largest which is an Int64
 end
+CLit(x) = CLit(CLit_tag(x), force_reinterpret(NTuple{8, UInt8}, x))
```

Comment:
What's an `NTuple`? Do you reinterpret the bytes of `x` as an 8-byte bit stream?

Reply:
`NTuple(8, UInt8)` is a short-hand for tuple of 8, `UInt8`s
i.e. `Tuple{UInt8,UInt8,UInt8,UInt8,UInt8,UInt8,UInt8,UInt8}`.
I just needed 64 bit of continuous memory to store the data.
It's type doesn't actually matter.
It occurs to me now, that rather than `NTuple{8, UInt8}` I could have just used a `UInt64`.
Or even declared a new primitive type (which would also work for the 192 byte case below)
`primitive type Payload64 64 end`
Calling it a  `NTuple{8, UInt8}` was helpful because when I was counting how many bytes i needed for the union.
But I think I will change to a custom primitive.

As to why: see this comment in the docstring for `TaggedUnion`
https://github.com/google-research/dex-lang/blob/b80d02d7c063a9207175d5afd98f1f16fa984907/julia/src/api_types.jl#L24-L27

Reply:
Great, thanks for the explanation. One sketchy thing in this solution is that it assumes that the byte ordering is consistent for both C and Julia, but I guess it's fine for now 🤷

Reply:
Isn't that promised by endianness?
So it should be fixed by architecture?
(unless you are on a rare mixed architecture machine, and choosing to used different things in your Julia and C compilers?)
Or does endianness/C not make any promises about were things endup if you give a pointer to data that is too big (I don't see how it could chose anywhere but the start)

Reply:
Yeah I think it's just important that Julia and the C layout rules agree. It's both a matter of endianness (the order the bytes of a value are written to memory), but also a matter of where the padding in a union is (if you write a 4B value into an 8B union, you could write it to the first half or the second half). It'd be good to think about it carefully, but at this point I'm happy with tests being green 🤷‍♂️

Reply:
Julia generally tries to match default C padding for structs, although, of course that's dependent on the C compiler, but in practice Clang matches GCC (intentionally), and that's what we do. Unlike C, which allows tweaking struct layout, Julia doesn't have any good tools for that (we'd probably add some low level hook and add macros that emit those hooks).

Reply:
Thanks for elaborating! It's a bit sad that C API is the standard FFI format given how underspecified it is 😕